### ISSUE TITLE
I can't use BBcode or format properly!

### ISSUE THREAD

**Issue by noeltheblank** [fchat-pidgin/fchat-pidgin#181]
Hello, I've been using the pidgin client a lot recently and it's pretty good! No problems, besides one.

When I'm typing I like to format my stuff but when I open the formatting menu, I can't choose any of the options. Why is this?

Is there any way for me to format my posts any other way than typing in `[b]Hello![/b]`?

**Comment by vobe** [comment]
Indeed, back in the dark times of F-Chat 1.0, there were no formatting buttons, everybody was used to typing the bbcode themselves.

It will never be possible to implement all of the f-list formatting features in pidgin's formatting menu, but implementing some basic formatting might not be too hard. I can take a shot at this.

**Comment by noeltheblank** [comment]
If you could, according to the formatting menu, can you please add bold, italics, underline, strikethrough sub and superscript, and colored text?

These are the options in the formatting menu, I'm not sure if icons could be added but I highly doubt it.

Also many thanks for trying to add the options, this was the only problem that caused me to switch back and forth from the official client to the pidgin one.

**Comment by LurkerHub** [comment]
The lack of formatting is really limiting.
If you could add it, it would be great!

**Comment by LurkerHub** [comment]
Revisiting.
Please add support for the basic formatting options that exist on Pidgin's "Font" menu

**Comment by vobe** [comment]
I won't be looking into this until the new protocol is out, because they may change the way formatting is handled. But I will accept patches if someone has the courage to try it

**Comment by LurkerHub** [comment]
> I won't be looking into this until the new protocol is out, because they may change the way formatting is handled. But I will accept patches if someone has the courage to risk having their changes broken in a few months

I am not familiar enough with the code (or Pidgin, or even GTK) to submit a proper patch, but I wrote a function to do a conversion [here](https://paste.ee/p/ouhUg).  If you can enable the formatting menu and call it before sending the message (right before calling `purple_markup_html_to_xhtml` I think, as: `message = flist_html_to_bbcode(message);`) it should take care of the `[b]`, `[i]`, `[u]`, `[s]` and `[url]` subset.

**Comment by vobe** [comment]
Note to self:

https://docs.pidgin.im/pidgin/2.x.y/connection_8h.html#a5073f95bd10eb9eed84fa58cbf1bb162

https://docs.pidgin.im/pidgin/2.x.y/conversation_8h.html#a28e64d2787ef92ddd7bc7295b66bf98d

**Comment by LurkerHub** [comment]
> Note to self:
> https://docs.pidgin.im/pidgin/2.x.y/connection_8h.html#a5073f95bd10eb9eed84fa58cbf1bb162
> https://docs.pidgin.im/pidgin/2.x.y/conversation_8h.html#a28e64d2787ef92ddd7bc7295b66bf98d

Note to vobe:
https://discord.com/invite/pB9fbVC

**Comment by LurkerHub** [comment]
> it should take care of the `[b]`, `[i]`, `[u]`, `[s]` and `[url]` subset.

I could probably hack something to work with `[color]` but the user would have to type the colour name into the picker (I checked, it works and converts to the RGB value), or possibly try to get the "nearest" colour.  Is it worth the effort?

`[sub]` and `[sup]` have no menu analogues.

**Comment by alexolog** [comment]
Hmmm...
I am not sure that the tags for some of the formats are correct.
Cannot test it though.

**Comment by LurkerHub** [comment]
Previous code was crap, I deleted it.

Looking at `tag_to_html_start()` and `tag_to_html_end()` in pidgin-2.14.4\pidgin\gtkimhtml.c my head a splode.  It seems that it can output both new and old-style tags?  Asking on the Discord didn't help.

I decided to support both, so the updated code can emit `[u]`, `[i]`, `[b]`, `[s]`, `[url]`, `[sup]`, `[sub]`, and `[color]` BBcode tags (and their corresponding closing tags), where the colour is approximated based on least square distance of the RGB components, and the `<sub>` and `<sup>` tags come from... somewhere?  They are present in the Pidgin source code I mentioned above so I handled them.

Anyway, here's the new code for your amusement: https://paste.ee/p/YPZuX

**Comment by LurkerHub** [comment]
@vobe can you integrate the code please and enable formatting?

**Comment by LurkerHub** [comment]
@vobe, can you please take a look?
I'd submit a proper pull-request, but I don't have an environment I can compile it on.
